Joint Graph Runner / AOT Eager Runner

SherlockNoMad opened the pull request:
After applying partitioner, you have fw_gm, and bw_gm. 

This PR implements a runner to run the graphs, via autograd.Function, providing the same calling convention as the original module. 

Sample output

fw_gm.py: P1971574036
bw_gm.py: P1971573886

you can modify fw_gm.py and bw_gm.py for fine grain customization! 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #164557
* #164373
* __->__ #164328



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk

SherlockNoMad committed:
codegen

[ghstack-poisoned]

vadimkantorov commented on test/distributed/tensor/test_dtensor_export.py:
```diff
@@ -150,6 +156,74 @@ def test_export_parallelize_module_with_dtensor_input(
+    @parametrize(
+        "run_mode", [RunMode.CODEGEN, RunMode.GRAPH_MODULE, RunMode.FX_INTERPRETER]
+    )
+    def test_jonit_graph_runner(self, run_mode):
```
typo "jonit" -> "joint"

vadimkantorov replied:
typo "jonit" -> "joint"

SherlockNoMad committed:
Update on "Joint Graph Runner / AOT Eager Runner "

After applying partitioner, you have fw_gm, and bw_gm. 

This PR implements a runner to run the graphs, via autograd.Function, providing the same calling convention as the original module. 

Sample output

fw_gm.py: P1971574036
bw_gm.py: P1971573886

you can modify fw_gm.py and bw_gm.py for fine grain customization! 




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

SherlockNoMad commented on torch/_functorch/_aot_autograd/aot_eager_runner.py:
```diff
@@ -0,0 +1,193 @@
+        # Run the forward graph module
+        # fw_graph_inputs = (*parameters, *buffers, *user_inputs)
+        # fw_graph_outputs = (*updated_inputs, *user_outputs, *saved_intermediates)
+        fw_outputs = fw_runner(*fw_args)
```
here

patrick-toulme commented:
Could you make an option to merge forward + backward into one GraphModule()? That would be ideal for performance. 

We can further extend to support a couple extensions
1. fwd+bwd+optimizer
2. gradient_accumultion * fwd + bwd + optimizer

All in one GraphModule

SherlockNoMad commented on torch/_functorch/_aot_autograd/aot_eager_runner.py:
```diff
@@ -0,0 +1,193 @@
+
+        # need a wrapper to pass the dummy None as the first argument `self`
+        def wrapped_fn(*args):
+            return fn(None, *args)
```
custom fx graph codegen here.
